test_trello_endtoend.py

```diff
@@ -0,0 +1,66 @@
+import os
+from threading import Thread
+
+import pytest
+import requests
+from selenium import webdriver
+from selenium.webdriver.common.keys import Keys
+
+import app
+from app import DEFAULT_PARAMS
+from trello_utils import TRELLO_BOARD, TRELLO_URL_BASE
+
+# TRELLO_KEY = os.environ.get('TRELLO_KEY')
+# TRELLO_TKN = os.environ.get('TRELLO_TKN')
+# DEFAULT_PARAMS = {'key': TRELLO_KEY, 'token': TRELLO_TKN}
+
+def create_board():
+    params = {'name': 'integration_test'}
+    params.update(DEFAULT_PARAMS)
+    print(DEFAULT_PARAMS)
+    print(TRELLO_URL_BASE)
+    response = requests.post(TRELLO_URL_BASE + 'boards/', data=params).json()
+    board_id = response['id']
+    return board_id
+
+
+def delete_board(board_id):
+    requests.delete(TRELLO_URL_BASE + 'boards/' + board_id, data=DEFAULT_PARAMS).json()
+
+
+@pytest.fixture(scope="module")
+def driver():
+    with webdriver.Firefox() as driver:
+        yield driver
+
+
+@pytest.fixture(scope='module')
+def test_app():
+    # Create the new board & update the board id environment variable
+    board_id = create_board()
+    os.environ['TRELLO_BOARD'] = board_id
+    # construct the new application
+    application = app.start_app()
+    # start the app in its own thread.
+    thread = Thread(target=lambda: application.run(use_reloader=False))
+    thread.daemon = True
+    thread.start()
+    yield app
+    # Tear Down
+    thread.join(1)
+    delete_board(board_id)
+
+
+def test_basic_endtoend_flow(driver, test_app):
+    driver.get('http://localhost:5000')
+    assert driver.title == 'To-Do App'
+
+
+def test_basic_create_card(driver, test_app):
+    driver.get('http://localhost:5000')
+    input = driver.find_element_by_id('new_card_textbox')
+    input.send_keys('test_card')
+    input.send_keys(Keys.ENTER)
+
+    driver.implicitly_wait(20)
```

Comment:
implicitly_wait means that the driver will up to that many seconds for an element to appear when you try to find one, e.g. with `find_element_by_id(...)`.

Which is what you should be doing - you should be trying to find the element. Looks like you even added `data-test-class="task"` so you should be able to do something with `find_elements_by_xpath("//*[@data-test-class='task']")`

What you are doing currently is immediately evaluating the expression `'test_card' in {some string}` and it is evaluating to false. (An explicit wait would probably get the test to pass but would result in a slower than necessary test)

Reply:
ohhhhhhhh i think i get it. Thanks.

Reply:
Not sure what you mean by `data-test-class="task"` though - I don't have that in my code.

Reply:
Done. Many thanks. All tests passing now.